PHP Version update

mmachatschek opened the pull request:
## Purpose

This fixes some PHP 7 related issues. #76 #77 #73 
Branch sniff was merged here with alle the changes to the new linter and also the fixes for the linter warnings.

## Approach

PHP 5.6 and PHP 7.0 support was offically dropped. As we want to provide software on the cutting edge, we are also going to drop the unsupported versions and take the chance to add all new language specific features into the code.

#### Open Questions and Pre-Merge TODOs

- [x] `composer lint` was executed.
- [x] Tests were written and pass with 100% coverage.
- [x] A issue with a detailed explanation of the problem/enhancement was created and linked.

mmachatschek committed:
Drop support for php 5.6 and 7.0 and update phpunit to current version (#74)

* Remove obsolete phpunit configuration
* Add assertions to tests which would raise an error with new version

Commit:
Merge branch 'develop' into php_version_update

mmachatschek committed:
Explicitly set constant visibility (#88)

mmachatschek committed:
Remove obsolet PHPUnit configuration entries (#82)

mmachatschek committed:
Add PHP 7.1 type declarations (#87)

howard committed:
Replace PHP CS Fixer with PHP Code Sniffer

Commit:
Fix code sniff warnings of too long line endings

Commit:
Move exceptions into seperate folder and update use statements

Commit:
Fix import statements and update tests

Commit:
Fix remaining codesniff errors

Commit:
Move pr template into github folder and update task

Commit:
Fix comment of code sniff ruleset

Commit:
Merge branch 'develop' into php_version_update

Commit:
Merge branch 'sniff' into php_version_update

mmachatschek committed:
Add return typing to all interfaces that can have them (#92)

* Set return types of the core
* Remove return types of optional elements for CSV export
* Add return types to tests
* Require PHP >= 7.1 and raise PHPUnit version

Commit:
Add php version 7.3 to travis configuration

howard commented on phpunit.xml:
```diff
@@ -2,14 +2,7 @@
 
 <phpunit
```
While we're here changing this file, we could add the [schema](https://schema.phpunit.de/7.5/phpunit.xsd) for this file so we can get IDE completion and warnings.

mmachatschek replied:
Done!

howard commented on src/FINDOLOGIC/Export/Data/Attribute.php:
```diff
@@ -53,7 +57,7 @@ public function getKey()
      * @inheritdoc
      */
-    public function getDomSubtree(\DOMDocument $document)
+    public function getDomSubtree(\DOMDocument $document): \DOMElement
```
Consider `use`ing `DOMElement`.

mmachatschek replied:
There will be a custom branch for this as soon we merged this into the develop branch. I also created a issue #97 for this.

howard commented on src/FINDOLOGIC/Export/Data/Property.php:
```diff
@@ -88,7 +77,8 @@ protected function setValues($values)
          */
         array_walk($values, function ($item, $key) {
```
Is it not possible to add typings to this closure?

mmachatschek replied:
I added type hintings to the closures in other classes where possible. For this one, we may not add the type hint because we would skip our specific warning from the check, if the provided array is associative. 

I'll have to leave it like that for now.

howard commented on src/FINDOLOGIC/Export/Exceptions/AttributeValueLengthException.php:
```diff
@@ -0,0 +1,15 @@
+
+class AttributeValueLengthException extends \RuntimeException
+{
+    public function __construct($attributeName, $characterLimit)
```
Typings?

mmachatschek replied:
Done!

howard commented on src/FINDOLOGIC/Export/Exceptions/BadPropertyKeyException.php:
```diff
@@ -0,0 +1,17 @@
+ */
+class BadPropertyKeyException extends \RuntimeException
+{
+    public function __construct($propertyKey)
```
Typing?

mmachatschek replied:
Done!

howard commented on src/FINDOLOGIC/Export/Exceptions/DuplicateValueForUsergroupException.php:
```diff
@@ -0,0 +1,11 @@
+
+class DuplicateValueForUsergroupException extends \RuntimeException
+{
+    public function __construct($key, $usergroup)
```
Typings?

mmachatschek replied:
Done!

howard commented on src/FINDOLOGIC/Export/Exceptions/EmptyElementsNotAllowedException.php:
```diff
@@ -0,0 +1,12 @@
+
+class EmptyElementsNotAllowedException extends \RuntimeException
+{
+    public function __construct($elementType, $elementKey)
```
Typings?

mmachatschek replied:
Done! Updated message too so we use a formated string here.

howard commented on src/FINDOLOGIC/Export/Exceptions/EmptyValueNotAllowedException.php:
```diff
@@ -0,0 +1,11 @@
+
+class EmptyValueNotAllowedException extends \RuntimeException
+{
+    public function __construct($message = 'Empty values are not allowed!')
```
Typings?

mmachatschek replied:
done!

howard commented on src/FINDOLOGIC/Export/Exceptions/PropertyKeyNotAllowedException.php:
```diff
@@ -0,0 +1,12 @@
+
+class PropertyKeyNotAllowedException extends \RuntimeException
+{
+    public function __construct($key)
```
Typings?

mmachatschek replied:
Done!

mmachatschek replied:
Done!

howard commented on src/FINDOLOGIC/Export/Exceptions/UnsupportedValueException.php:
```diff
@@ -0,0 +1,14 @@
+
+class UnsupportedValueException extends \BadMethodCallException
+{
+    public function __construct($unsupportedValueName)
```
Typings?

mmachatschek replied:
Done!

howard commented on src/FINDOLOGIC/Export/Exceptions/ValueIsNotIntegerException.php:
```diff
@@ -0,0 +1,11 @@
+
+class ValueIsNotIntegerException extends \RuntimeException
+{
+    public function __construct($value)
```
Typings?

mmachatschek replied:
Done!

howard commented on src/FINDOLOGIC/Export/Exceptions/ValueIsNotPositiveIntegerException.php:
```diff
@@ -0,0 +1,11 @@
+
+class ValueIsNotPositiveIntegerException extends \RuntimeException
+{
+    public function __construct($value)
```
Typings?

mmachatschek replied:
Done!

howard reviewed (changes requested):
* Just some minor stuff, like possibly missing typings.
* Can we teach the linter to criticize missing typings to force us to explain what we're doing in a comment in case we can't specify a type?

Commit:
Update composer file and load most recent PHPUnit version

Commit:
Add phpunit version 8 cache file to ignored files

Commit:
Fix fatal errors with phpunit version 8

Commit:
Remove expectedException annotations from docblock in tests

Commit:
Update travis config to use build stages

Commit:
Update phpunit and composer configuration

Commit:
Merge branch 'develop' into php_version_update

Commit:
Add schema definition to phpunit configuration file

Commit:
Remove symfony console from composer file and update codesniffer

Commit:
Rename codesniffer configuration file

Commit:
Add missing type hints

Commit:
Use formated string in exception message

mmachatschek commented:
@howard 
> Can we teach the linter to criticize missing typings to force us to explain what we're doing in a comment in case we can't specify a type?

I searched in the codesniffer project as good as I can but I didn't find the right rule/ruleset for return types. Maybe you know of one?

Furthermore I also updated the composer and travis configs again.

Composer now uses a more frequent version of phpunit and codesniffer and also gets the most current version for each PHP version. The symfony console dependency was also removed because there is no usage for it. The parameters from the `composer test` command were moved to the phpunit config file.

Travis build is also staged now like we have it in the [shopware project](https://github.com/findologic/plugin-shopware-unified)

howard commented:
> I searched in the codesniffer project as good as I can but I didn't find the right rule/ruleset for return types. Maybe you know of one?

I asked because I couldn't find one myself, so let's leave it at that. Maybe we can try implementing our own sniffs in the future.

>Travis build is also staged now like we have it in the shopware project

I have mixed feelings about staged builds. They are great for isolation, but for each stage, VM provisioning and setup tasks need to be repeated, adding to build time. Normally, anyway. Since this just seems to add one quick stage and dependency install is quick (other than certain projects that have to compile PHP extensions), plus old PHP version stages are removed, it's overall quicker. So, nice!

Commit:
Merge branch 'develop' into php_version_update

Commit:
Remove stages from travis build

mmachatschek commented:
@howard removed staged build again. There should be only three testruns now, performance wise this is the quickest for now.

howard commented on phpcs.xml.dist:
```diff
@@ -0,0 +1,26 @@
+<?xml version="1.0"?>
```
I don't understand why this should be a `.dist` file. Such files are only relevant if local changes on a copy without `.dist` in the file name are required for development or production. Either is not the case here, as the checked in version should *always* be used.

mmachatschek replied:
The advantage here is that the none `.dist` files are loaded before the `.dist` files which has the advantage that I can add extra configuration like a PHPUnit HTML output for development purpose. I usally do this to check the coverage of my code and that way I don't need to run a custom command and just add the phpunit.xml file to the project and everything works smooth.

In the .gitignore file the `phpunit.xml` file was added so it doesnt get commited by mistake.

howard replied:
Didn't know about this mechanism. Nifty! Consider me convinced.

howard commented on phpunit.xml.dist:
```diff
@@ -0,0 +1,26 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
I don't understand why this should be a `.dist` file. Such files are only relevant if local changes on a copy without `.dist` in the file name are required for development or production. Either is not the case here, as the checked in version should *always* be used.

mmachatschek replied:
The advantage here is that the none `.dist` files are loaded before the `.dist` files which has the advantage that I can add extra configuration like a PHPUnit HTML output for development purpose. I usally do this to check the coverage of my code and that way I don't need to run a custom command and just add the phpunit.xml file to the project and everything works smooth.

In the .gitignore file the `phpunit.xml` file was added so it doesnt get commited by mistake.

howard commented on phpcs.xml.dist:
```diff
@@ -0,0 +1,26 @@
+<?xml version="1.0"?>
+<ruleset name="FINDOLOGIC">
```
There is a [schema](https://github.com/squizlabs/PHP_CodeSniffer/blob/3.4.1/phpcs.xsd) for this file. We should use it to enable IDE support.

mmachatschek replied:
Added schema

howard commented on tests/FINDOLOGIC/Export/Tests/XmlSerializationTest.php:
```diff
@@ -30,19 +35,20 @@
 class XmlSerializationTest extends TestCase
 {
-    const SCHEMA_URL = 'https://raw.githubusercontent.com/findologic/xml-export/master/src/main/resources/findologic.xsd';
+    private const SCHEMA_URL = 'https://raw.githubusercontent.com/findologic/xml-export/master/src/main/resources/' .
```
Insead of downloading the schema like this, use [findologic/xml-export-schema](https://packagist.org/packages/findologic/xml-export-schema) as a dev dependency.

mmachatschek replied:
Added the class `Constant` from the project. We just need to wait for the packagist update of the dependency with version `1.2.0`.

howard replied:
@mmachatschek The new version is now on Packagist.

howard reviewed (changes requested):
Only some minor suggested changes. Nice work! I'm looking forward to release 2.0!

Commit:
Add schema link to code sniffer config file

Commit:
Load the xsd schema via package not via http

howard committed:
Add maintenance information to the README

howard committed:
Update contribution guide
